File: cpp/src/mq/pulsar.cpp
#include "mental1104/mq/pulsar.h"
#include "mental1104/mq/transport.h"

#include <mutex>
#include <stdexcept>

#ifdef M1104_HAS_PULSAR
#include <pulsar/Client.h>
#include <pulsar/Consumer.h>
#include <pulsar/Message.h>
#include <pulsar/Producer.h>
#endif

namespace mental1104 {
namespace mq {

#ifdef M1104_HAS_PULSAR
namespace {

pulsar::ConsumerType to_pulsar_type(SubscriptionType type) {
  switch (type) {
  case SubscriptionType::Exclusive:
    return pulsar::ConsumerExclusive;
  case SubscriptionType::Failover:
    return pulsar::ConsumerFailover;
  case SubscriptionType::KeyShared:
    return pulsar::ConsumerKeyShared;
  case SubscriptionType::Shared:
  default:
    return pulsar::ConsumerShared;
  }
}

class PulsarProducerTransport : public ProducerTransport {
public:
  explicit PulsarProducerTransport(const pulsar::Producer &producer)
      : producer_(producer), closed_(false) {}

  SendResult send(const Record &record) override {
    ensure_open();
    pulsar::Message message =
        pulsar::MessageBuilder()
            .setContent(record.empty()
                            ? ""
                            : reinterpret_cast<const char *>(&record[0]),
                        record.size())
            .build();
    pulsar::MessageId id;
    const pulsar::Result result = producer_.send(message, id);
    if (result != pulsar::ResultOk) {
      return SendResult::failure(pulsar::strResult(result));
    }
    return SendResult::success(id.serialize());
  }

  void send_async(const Record &record, const SendCallback &callback) override {
    ensure_open();
    pulsar::Message message =
        pulsar::MessageBuilder()
            .setContent(record.empty()
                            ? ""
                            : reinterpret_cast<const char *>(&record[0]),
                        record.size())
            .build();
    producer_.sendAsync(message,
                        [callback](pulsar::Result result,
                                   const pulsar::MessageId &id) {
                          if (!callback) {
                            return;
                          }
                          if (result == pulsar::ResultOk) {
                            callback(SendResult::success(id.serialize()));
                          } else {
                            callback(SendResult::failure(
                                pulsar::strResult(result)));
                          }
                        });
  }

  void close() override {
    std::lock_guard<std::mutex> lock(mutex_);
    if (closed_) {
      return;
    }
    closed_ = true;
    producer_.flush();
    producer_.close();
  }

private:
  void ensure_open() const {
    if (closed_) {
      throw std::runtime_error("Pulsar producer is closed");
    }
  }

  pulsar::Producer producer_;
  mutable std::mutex mutex_;
  bool closed_;
};

class PulsarMessageHandle : public MessageHandle {
public:
  explicit PulsarMessageHandle(const pulsar::Message &message) : message(message) {}
  pulsar::Message message;
};

class PulsarConsumerTransport : public ConsumerTransport {
public:
  PulsarConsumerTransport(pulsar::Client &client, const std::string &topic,
                          const std::string &subscription,
                          const pulsar::ConsumerConfiguration &configuration)
      : client_(client), topic_(topic), subscription_(subscription),
        configuration_(configuration), closed_(false) {
    subscribe_new();
  }

  MessagePtr receive(int timeout_millis) override {
    ensure_open();
    pulsar::Message native;
    const pulsar::Result result =
        timeout_millis < 0 ? consumer_.receive(native)
                           : consumer_.receive(native, timeout_millis);
    if (result == pulsar::ResultTimeout) {
      throw std::runtime_error("message receive timed out");
    }
    if (result != pulsar::ResultOk) {
      throw std::runtime_error("Pulsar receive: " +
                               pulsar::strResult(result));
    }
    MessagePtr message(new Message());
    const std::string payload = native.getDataAsString();
    message->payload.assign(payload.begin(), payload.end());
    message->native.reset(new PulsarMessageHandle(native));
    return message;
  }

  void acknowledge(const MessagePtr &message) override {
    const pulsar::Result result =
        consumer_.acknowledge(handle_of(message)->message);
    if (result != pulsar::ResultOk) {
      throw std::runtime_error("Pulsar acknowledge: " +
                               pulsar::strResult(result));
    }
  }

  void negative_acknowledge(const MessagePtr &message) override {
    consumer_.negativeAcknowledge(handle_of(message)->message);
  }

  void unsubscribe() override {
    ensure_open();
    const pulsar::Result result = consumer_.unsubscribe();
    if (result != pulsar::ResultOk) {
      throw std::runtime_error("Pulsar unsubscribe: " +
                               pulsar::strResult(result));
    }
  }

  void resubscribe() override {
    ensure_open();
    consumer_.close();
    subscribe_new();
  }

  void close() override {
    if (closed_) {
      return;
    }
    closed_ = true;
    consumer_.close();
  }

private:
  void ensure_open() const {
    if (closed_) {
      throw std::runtime_error("Pulsar consumer is closed");
    }
  }

  std::shared_ptr<PulsarMessageHandle>
  handle_of(const MessagePtr &message) const {
    if (!message || !message->native) {
      throw std::invalid_argument("message does not belong to Pulsar consumer");
    }
    std::shared_ptr<PulsarMessageHandle> handle =
        std::dynamic_pointer_cast<PulsarMessageHandle>(message->native);
    if (!handle) {
      throw std::invalid_argument("message does not belong to Pulsar consumer");
    }
    return handle;
  }

  void subscribe_new() {
    const pulsar::Result result =
        client_.subscribe(topic_, subscription_, configuration_, consumer_);
    if (result != pulsar::ResultOk) {
      throw std::runtime_error("Pulsar subscribe: " +
                               pulsar::strResult(result));
    }
  }

  pulsar::Client &client_;
  std::string topic_;
  std::string subscription_;
  pulsar::ConsumerConfiguration configuration_;
  pulsar::Consumer consumer_;
  bool closed_;
};

} // namespace
#endif

class PulsarMessageQueue::Impl {
public:
  explicit Impl(const Options &config)
#ifdef M1104_HAS_PULSAR
      : client(option(config, "service.url")), config(config), closed(false) {}
#else
      : config(config), closed(false) {}
#endif

#ifdef M1104_HAS_PULSAR
  static std::string option(const Options &config, const std::string &key) {
    Options::const_iterator it = config.find(key);
    if (it == config.end() || it->second.empty()) {
      throw std::invalid_argument("missing Pulsar option: " + key);
    }
    return it->second;
  }
  pulsar::Client client;
#endif
  Options config;
  bool closed;
  std::mutex mutex;
};

bool pulsar_available() {
#ifdef M1104_HAS_PULSAR
  return true;
#else
  return false;
#endif
}

PulsarMessageQueue::PulsarMessageQueue(const Options &config)
    : impl_(new Impl(config)) {}

PulsarMessageQueue::~PulsarMessageQueue() { close(); }

std::shared_ptr<AbstractProducer> PulsarMessageQueue::create_producer(
    const std::string &tenant, const std::string &namespace_name,
    const std::string &topic, const Schema &schema, bool batching_enabled) {
  (void)schema;
  const std::string full_topic =
      build_pulsar_topic(tenant, namespace_name, topic);
  std::lock_guard<std::mutex> lock(impl_->mutex);
  if (impl_->closed) {
    throw std::runtime_error("PulsarMessageQueue is closed");
  }
#ifdef M1104_HAS_PULSAR
  pulsar::ProducerConfiguration configuration;
  configuration.setBatchingEnabled(batching_enabled);
  pulsar::Producer native;
  const pulsar::Result result =
      impl_->client.createProducer(full_topic, configuration, native);
  if (result != pulsar::ResultOk) {
    throw std::runtime_error("create Pulsar producer: " +
                             pulsar::strResult(result));
  }
  return std::shared_ptr<AbstractProducer>(new Producer(
      std::shared_ptr<ProducerTransport>(new PulsarProducerTransport(native))));
#else
  (void)full_topic;
  (void)batching_enabled;
  throw std::runtime_error(
      "Pulsar support is unavailable; install pulsar-client-cpp headers and library");
#endif
}

std::shared_ptr<AbstractConsumer> PulsarMessageQueue::create_consumer(
    const std::string &tenant, const std::string &namespace_name,
    const std::string &topic, const std::string &subscription,
    const Schema &schema, SubscriptionType subscription_type,
    const MessageListener &message_listener, const Options &options) {
  (void)schema;
  (void)options;
  if (subscription.empty()) {
    throw std::invalid_argument("Pulsar subscription must not be empty");
  }
  const std::string full_topic =
      build_pulsar_topic(tenant, namespace_name, topic);
  std::lock_guard<std::mutex> lock(impl_->mutex);
  if (impl_->closed) {
    throw std::runtime_error("PulsarMessageQueue is closed");
  }
#ifdef M1104_HAS_PULSAR
  pulsar::ConsumerConfiguration configuration;
  configuration.setConsumerType(to_pulsar_type(subscription_type));
  return std::shared_ptr<AbstractConsumer>(new Consumer(
      std::shared_ptr<ConsumerTransport>(new PulsarConsumerTransport(
          impl_->client, full_topic, subscription, configuration)),
      message_listener));
#else
  (void)full_topic;
  (void)subscription_type;
  (void)message_listener;
  throw std::runtime_error(
      "Pulsar support is unavailable; install pulsar-client-cpp headers and library");
#endif
}

void PulsarMessageQueue::close() {
  std::lock_guard<std::mutex> lock(impl_->mutex);
  if (impl_->closed) {
    return;
  }
  impl_->closed = true;
#ifdef M1104_HAS_PULSAR
  impl_->client.close();
#endif
}

} // namespace mq
} // namespace mental1104

File: golang/mental1104/mq/pulsar/pulsar.go
package pulsar

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"time"

	commonmq "github.com/mental1104/common/golang/mental1104/mq"
	pulsargo "github.com/apache/pulsar-client-go/pulsar"
)

type Config struct {
	ClientOptions  pulsargo.ClientOptions
	ProducerConfig func(*pulsargo.ProducerOptions)
	ConsumerConfig func(*pulsargo.ConsumerOptions)
	CloseTimeout   time.Duration
}

type client interface {
	CreateProducer(pulsargo.ProducerOptions) (producerClient, error)
	Subscribe(pulsargo.ConsumerOptions) (consumerClient, error)
	Close()
}

type nativeClient struct{ pulsargo.Client }

func (c nativeClient) CreateProducer(options pulsargo.ProducerOptions) (producerClient, error) {
	return c.Client.CreateProducer(options)
}
func (c nativeClient) Subscribe(options pulsargo.ConsumerOptions) (consumerClient, error) {
	return c.Client.Subscribe(options)
}

type producerClient interface {
	Send(context.Context, *pulsargo.ProducerMessage) (pulsargo.MessageID, error)
	SendAsync(context.Context, *pulsargo.ProducerMessage, func(pulsargo.MessageID, *pulsargo.ProducerMessage, error))
	FlushWithCtx(context.Context) error
	Close()
}

type consumerClient interface {
	Receive(context.Context) (pulsargo.Message, error)
	Ack(pulsargo.Message) error
	Nack(pulsargo.Message)
	Unsubscribe() error
	Close()
}

type MessageQueue struct {
	mu             sync.Mutex
	closed         bool
	client         client
	ownsClient     bool
	producerConfig func(*pulsargo.ProducerOptions)
	consumerConfig func(*pulsargo.ConsumerOptions)
	closeTimeout   time.Duration
}

type PulsarMessageQueue = MessageQueue
type PulsarProducer = Producer
type PulsarConsumer = Consumer

func NewMessageQueue(config Config) (*MessageQueue, error) {
	if config.ClientOptions.URL == "" {
		return nil, errors.New("pulsar client URL must not be empty")
	}
	native, err := pulsargo.NewClient(config.ClientOptions)
	if err != nil {
		return nil, fmt.Errorf("create pulsar client: %w", err)
	}
	q := newMessageQueue(nativeClient{Client: native}, config)
	q.ownsClient = true
	return q, nil
}

func newMessageQueue(native client, config Config) *MessageQueue {
	timeout := config.CloseTimeout
	if timeout <= 0 {
		timeout = 10 * time.Second
	}
	return &MessageQueue{
		client:         native,
		producerConfig: config.ProducerConfig,
		consumerConfig: config.ConsumerConfig,
		closeTimeout:   timeout,
	}
}

func (q *MessageQueue) CreateProducer(_ context.Context, tenant, namespace, topic string, _ commonmq.Schema, batchingEnabled bool) (commonmq.AbstractProducer, error) {
	fullTopic, err := commonmq.BuildPulsarTopic(tenant, namespace, topic)
	if err != nil {
		return nil, err
	}
	q.mu.Lock()
	defer q.mu.Unlock()
	if q.closed {
		return nil, commonmq.ErrClosed
	}
	options := pulsargo.ProducerOptions{Topic: fullTopic, DisableBatching: !batchingEnabled}
	if q.producerConfig != nil {
		q.producerConfig(&options)
	}
	native, err := q.client.CreateProducer(options)
	if err != nil {
		return nil, fmt.Errorf("create pulsar producer: %w", err)
	}
	return &Producer{producer: native, closeTimeout: q.closeTimeout}, nil
}

func (q *MessageQueue) CreateConsumer(_ context.Context, tenant, namespace, topic, subscription string, _ commonmq.Schema, options commonmq.ConsumerOptions) (commonmq.AbstractConsumer, error) {
	fullTopic, err := commonmq.BuildPulsarTopic(tenant, namespace, topic)
	if err != nil {
		return nil, err
	}
	if subscription == "" {
		return nil, errors.New("pulsar subscription must not be empty")
	}
	q.mu.Lock()
	defer q.mu.Unlock()
	if q.closed {
		return nil, commonmq.ErrClosed
	}
	subscriptionType := pulsargo.Shared
	if value, ok := options.SubscriptionType.(pulsargo.SubscriptionType); ok {
		subscriptionType = value
	}
	nativeOptions := pulsargo.ConsumerOptions{
		Topic:            fullTopic,
		SubscriptionName: subscription,
		Type:             subscriptionType,
	}
	if q.consumerConfig != nil {
		q.consumerConfig(&nativeOptions)
	}
	native, err := q.client.Subscribe(nativeOptions)
	if err != nil {
		return nil, fmt.Errorf("create pulsar consumer: %w", err)
	}
	return &Consumer{
		client:   q.client,
		consumer: native,
		options:  nativeOptions,
		listener: options.MessageListener,
	}, nil
}

func (q *MessageQueue) Close() error {
	q.mu.Lock()
	if q.closed {
		q.mu.Unlock()
		return nil
	}
	q.closed = true
	owns := q.ownsClient
	native := q.client
	q.mu.Unlock()
	if owns && native != nil {
		native.Close()
	}
	return nil
}

type Producer struct {
	mu           sync.Mutex
	closed       bool
	wg           sync.WaitGroup
	producer     producerClient
	closeTimeout time.Duration
}

func (p *Producer) begin() error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.closed {
		return commonmq.ErrClosed
	}
	p.wg.Add(1)
	return nil
}

func (p *Producer) Send(ctx context.Context, record any) error {
	if err := p.begin(); err != nil {
		return err
	}
	defer p.wg.Done()
	payload, err := commonmq.MarshalRecord(record)
	if err != nil {
		return err
	}
	_, err = p.producer.Send(ctx, &pulsargo.ProducerMessage{Payload: payload})
	if err != nil {
		return fmt.Errorf("pulsar send: %w", err)
	}
	return nil
}

func (p *Producer) SendAsync(ctx context.Context, record any, callback commonmq.SendCallback) error {
	if err := p.begin(); err != nil {
		return err
	}
	payload, err := commonmq.MarshalRecord(record)
	if err != nil {
		p.wg.Done()
		return err
	}
	message := &pulsargo.ProducerMessage{Payload: payload}
	p.producer.SendAsync(ctx, message, func(id pulsargo.MessageID, _ *pulsargo.ProducerMessage, sendErr error) {
		if sendErr != nil {
			sendErr = fmt.Errorf("pulsar async send: %w", sendErr)
		}
		messageID := ""
		if id != nil {
			messageID = id.String()
		}
		p.wg.Done()
		if callback != nil {
			callback(commonmq.SendResult{MessageID: messageID, Err: sendErr})
		}
	})
	return nil
}

func (p *Producer) Close() error {
	p.mu.Lock()
	if p.closed {
		p.mu.Unlock()
		return nil
	}
	p.closed = true
	p.mu.Unlock()
	ctx, cancel := context.WithTimeout(context.Background(), p.closeTimeout)
	flushErr := p.producer.FlushWithCtx(ctx)
	cancel()
	p.producer.Close()
	p.wg.Wait()
	return flushErr
}

type Consumer struct {
	opMu     sync.Mutex
	closed   bool
	client   client
	consumer consumerClient
	options  pulsargo.ConsumerOptions
	listener commonmq.MessageListener
}

func (c *Consumer) Receive(ctx context.Context, timeout time.Duration) (*commonmq.Message, error) {
	c.opMu.Lock()
	if c.closed || c.consumer == nil {
		c.opMu.Unlock()
		return nil, commonmq.ErrClosed
	}
	if timeout > 0 {
		var cancel context.CancelFunc
		ctx, cancel = context.WithTimeout(ctx, timeout)
		defer cancel()
	}
	native, err := c.consumer.Receive(ctx)
	if err != nil {
		c.opMu.Unlock()
		if errors.Is(err, context.DeadlineExceeded) {
			return nil, commonmq.ErrTimeout
		}
		return nil, fmt.Errorf("pulsar receive: %w", err)
	}
	message := &commonmq.Message{Payload: append([]byte(nil), native.Payload()...), Native: native}
	listener := c.listener
	c.opMu.Unlock()
	if listener != nil {
		listener(message)
	}
	return message, nil
}

func pulsarMessage(message *commonmq.Message) (pulsargo.Message, error) {
	if message == nil {
		return nil, commonmq.ErrInvalidMessage
	}
	native, ok := message.Native.(pulsargo.Message)
	if !ok || native == nil {
		return nil, commonmq.ErrInvalidMessage
	}
	return native, nil
}

func (c *Consumer) Acknowledge(_ context.Context, message *commonmq.Message) error {
	native, err := pulsarMessage(message)
	if err != nil {
		return err
	}
	c.opMu.Lock()
	defer c.opMu.Unlock()
	if c.closed || c.consumer == nil {
		return commonmq.ErrClosed
	}
	if err := c.consumer.Ack(native); err != nil {
		return fmt.Errorf("pulsar acknowledge: %w", err)
	}
	return nil
}

func (c *Consumer) NegativeAcknowledge(_ context.Context, message *commonmq.Message) error {
	native, err := pulsarMessage(message)
	if err != nil {
		return err
	}
	c.opMu.Lock()
	defer c.opMu.Unlock()
	if c.closed || c.consumer == nil {
		return commonmq.ErrClosed
	}
	c.consumer.Nack(native)
	return nil
}

func (c *Consumer) Unsubscribe(_ context.Context) error {
	c.opMu.Lock()
	defer c.opMu.Unlock()
	if c.closed {
		return commonmq.ErrClosed
	}
	if c.consumer == nil {
		return nil
	}
	if err := c.consumer.Unsubscribe(); err != nil {
		return fmt.Errorf("pulsar unsubscribe: %w", err)
	}
	c.consumer = nil
	return nil
}

func (c *Consumer) Resubscribe(_ context.Context) error {
	c.opMu.Lock()
	defer c.opMu.Unlock()
	if c.closed {
		return commonmq.ErrClosed
	}
	if c.consumer != nil {
		c.consumer.Close()
	}
	native, err := c.client.Subscribe(c.options)
	if err != nil {
		c.consumer = nil
		return fmt.Errorf("pulsar resubscribe: %w", err)
	}
	c.consumer = native
	return nil
}

func (c *Consumer) Close() error {
	c.opMu.Lock()
	defer c.opMu.Unlock()
	if c.closed {
		return nil
	}
	c.closed = true
	if c.consumer != nil {
		c.consumer.Close()
		c.consumer = nil
	}
	return nil
}
